library/net.c

```diff
@@ -714,7 +713,7 @@ int dnet_process_recv(struct dnet_backend_io *backend, struct dnet_net_state *st
 			 */
 
 			pthread_mutex_lock(&st->trans_lock);
-			dnet_trans_timestamp(st, t);
+			dnet_trans_insert_timer_nolock(st, t);
```

Comment:
This particular chunk updates timestamp after transaction has been processed. I agree, this has to be saved, since callback may take a lot of time to complete, and after transaction has been put back, it can be immediately killed with timeout, I've updated pull request.

If we update timestamp in the network thread, it will behave the same way as currently - without updating it after completion in IO thread, checking thread may grab it and free. I think that updating timestamp in network thread instead of (or better in addition to) updating it in line 666, i.e. prior calling completion callback, might be a good idea.

I will update this patch and test this idea